In [20]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests

In [21]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

## Extract data of Toronto neighborhoods from Wikipedia

In [22]:
text_result = requests.get(url).text #get the entire html of the article as a str
html_parsed_result = BeautifulSoup(text_result, 'html.parser') #transform the text to html

neightborhood_info_table = html_parsed_result.find('table', class_ = 'wikitable')
neightborhood_rows = neightborhood_info_table.find_all('tr')

# extract the info ('Postcode', 'Borough', 'Neighbourhood') from the table
neightborhood_info = []
for row in neightborhood_rows:
    info = row.text.split('\n')[1:-1] # remove empty str (first and last items)
    neightborhood_info.append(info)
    
neightborhood_info[0:10]

[['Postcode', 'Borough', 'Neighbourhood'],
 ['M1A', 'Not assigned', 'Not assigned'],
 ['M2A', 'Not assigned', 'Not assigned'],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Harbourfront'],
 ['M5A', 'Downtown Toronto', 'Regent Park'],
 ['M6A', 'North York', 'Lawrence Heights'],
 ['M6A', 'North York', 'Lawrence Manor'],
 ['M7A', "Queen's Park", 'Not assigned']]

## Transform data into pandas dataframe

In [23]:
#create a Neighborhoods dataframe
neightborhood_info[0][-1] = 'Neighborhood' # change to american spelling
neighborhood_df = pd.DataFrame(neightborhood_info[1:], columns=neightborhood_info[0])

neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights
7      M6A        North York    Lawrence Manor
8      M7A      Queen's Park      Not assigned
9      M8A      Not assigned      Not assigned

## Begin the process to clean data  and get rid of 'Not assigned' values

In [24]:
not_assigned_boroughs = neighborhood_df.index[neighborhood_df['Borough'] == 'Not assigned']
not_assigned_neighborhoods = neighborhood_df.index[neighborhood_df['Neighborhood'] == 'Not assigned']
not_assigned_neighborhoods_and_borough = not_assigned_boroughs & not_assigned_neighborhoods

print('The DataFrame shape is {}'.format(neighborhood_df.shape),'\n')
print('There are:')
print('  {} Postal codes'.format(neighborhood_df['Postcode'].unique().shape[0]))
print('  {} Boroughs'.format(neighborhood_df['Borough'].unique().shape[0] - 1)) # substract one because "not assigned" doesn't count
print('  {} Neighborhoods'.format(neighborhood_df['Neighborhood'].unique().shape[0] - 1)) # substract one because "not assigned" doesn't count
print('  {} rows with Not assigned Borough'.format(not_assigned_boroughs.shape[0]))
print('  {} rows with Not assigned Neighborhood'.format(not_assigned_neighborhoods.shape[0]))
print('  {} rows with Not assigned Neighborhood and Borough'.format(not_assigned_neighborhoods_and_borough.shape[0]),'\n')

The DataFrame shape is (289, 3) 

There are:
  180 Postal codes
  11 Boroughs
  209 Neighborhoods
  77 rows with Not assigned Borough
  78 rows with Not assigned Neighborhood
  77 rows with Not assigned Neighborhood and Borough 



## Eliminate rows with a 'Not assigned' value for Borough

In [25]:
neighborhood_df.drop(neighborhood_df.index[not_assigned_boroughs], inplace=True)
neighborhood_df.reset_index(drop=True, inplace=True)

neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Not assigned
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

## The 'Not assigned' values in the Neighborhood column will be replace with the Borough name in that cell

In [26]:
not_assigned_neighborhoods = neighborhood_df.index[neighborhood_df['Neighborhood'] == 'Not assigned'] # run this again because the indexes on the dataframe where reset

for idx in not_assigned_neighborhoods:
    neighborhood_df['Neighborhood'][idx] = neighborhood_df['Borough'][idx]
    
neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights
5      M6A        North York    Lawrence Manor
6      M7A      Queen's Park      Queen's Park
7      M9A         Etobicoke  Islington Avenue
8      M1B       Scarborough             Rouge
9      M1B       Scarborough           Malvern

## Group the data by postal code since there are multiple rows for some postal codes

In [27]:
group = neighborhood_df.groupby('Postcode')
grouped_neighborhoods = group['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
grouped_boroughs = group['Borough'].apply(lambda x: set(x).pop())
grouped_df = pd.DataFrame(list(zip(grouped_boroughs.index, grouped_boroughs, grouped_neighborhoods)))
grouped_df.columns = ['Postcode', 'Borough', 'Neighborhood']

grouped_df.head(10)

Postcode      Borough                                     Neighborhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

In [45]:
grouped_df.rename(columns={'Postcode': 'Postal Code'}, inplace=True)

In [46]:
grouped_df.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [47]:
print('The DataFrame shape is', grouped_df.shape)

The DataFrame shape is (103, 3)


### There are 103 unique postal codes in the dataframe

In [48]:
coordinates_df = pd.read_csv('Geospatial_Coordinates.csv') # transform the csv file into a dataframe

print('The coordinates dataframe shape is', coordinates_df.shape)
coordinates_df.head()

The coordinates dataframe shape is (103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

## Add coordinates to the neighborhoods dataframe by joining them using the Postal Code as keys

In [49]:
postcodes_with_coordinates_df = grouped_df.join(coordinates_df.set_index('Postal Code'), on='Postal Code')

postcodes_with_coordinates_df.head(16)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   

                                         Neighborhood   Latitude  Longitude  
0                                      Rouge, Malvern  43.806686 -79.194353  
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8     Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                      L'Amoreaux West, Steeles West  43.799525 -79.318389

In [52]:
!pip install folium
import folium
map = folium.Map(location=[43.6532,-79.3832], zoom_start=11)

for location in postcodes_with_coordinates_df.itertuples(): #iterate each row of the dataframe
    label = 'Postal Code: {};  Borough: {};  Neighborhoods: {}'.format(location[1], location[2], location[3])
    label = folium.Popup(label, parse_html=True)    
    folium.CircleMarker(
        [location[-2], location[-1]],
        radius=1,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map) 
    folium.Circle(
        radius=500,
        popup=label,
        location=[location[-2], location[-1]],
        color='#3186cc',
        fill=True,
        fill_color='#3186cc'
    ).add_to(map) 
    
map

    100% |████████████████████████████████| 92kB 9.3MB/s 
You are using pip version 19.0.1, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


## We enter our Foursquare credentials.

In [54]:
CLIENT_ID = 'PWAJT2QWLEAIHWB4AKCAKEQ1VXOM0W4DGT0DIIM4VHMW5ERL' # your Foursquare ID
CLIENT_SECRET = 'CSZ4MLCYIWKNBF5DH5DIJBCKSSWKX0YNKZQXHAQGYFHE0GL4' # your Foursquare Secret
VERSION = '20181102'

In [59]:
Scarborough_data = postcodes_with_coordinates_df[postcodes_with_coordinates_df['Borough'] == 'Scarborough'].reset_index(drop=True)
Scarborough_data

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   
16         M1X  Scarborough   

                                         Neighborhood   Latitude  Longitude  
0                                      Rouge, Malvern  43.806686 -79.194353  
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8     Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                      L'Amoreaux West, Steeles West  43.799525 -79.318389  
16                                        Upper Rouge  43.836125 -79.205636

## For this analysis we will use the Borough of 'Scarborough' so we find the coordinates.

In [61]:
address = 'Scarborough, CA'

#geolocator = Nominatim()
#location = geolocator.geocode(address)
#latitude = location.latitude
#longitude = location.longitude
latitude = 43.773077
longitude = -79.257774
print('The geograpical coordinates of Scarborough are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Scarborough are 43.773077, -79.257774.


## Next, we use folium to create a map of Scarborough.

In [63]:
# create map of Scarborough using latitude and longitude values
map_Scarborough = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Scarborough_data['Latitude'], Scarborough_data['Longitude'], Scarborough_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_Scarborough)  
    
map_Scarborough

In [66]:
Scarborough_data.loc[0, 'Neighborhood']

'Rouge, Malvern'

In [67]:
Scarborough_data.shape

(17, 5)

In [68]:
neighborhood_latitude = Scarborough_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Scarborough_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = Scarborough_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


## Next, we get the necessary data from Foursquare.

In [69]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=PWAJT2QWLEAIHWB4AKCAKEQ1VXOM0W4DGT0DIIM4VHMW5ERL&client_secret=CSZ4MLCYIWKNBF5DH5DIJBCKSSWKX0YNKZQXHAQGYFHE0GL4&v=20181102&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

## After that, we get the json from the url.

In [70]:

s = requests.get(url)
results = s.json()

## We then utilize a function to get the category type.

In [71]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [72]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy's  Fast Food Restaurant  43.807448 -79.199056

## Then we use a function to identify the nearby venues.

In [73]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [74]:

Scarborough_venues = getNearbyVenues(names=Scarborough_data['Neighborhood'],
                                   latitudes=Scarborough_data['Latitude'],
                                   longitudes=Scarborough_data['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West, Steeles West
Upper Rouge


In [75]:
Scarborough_data.drop(index=16,axis=0,inplace=True)
Scarborough_data.tail()

Postal Code      Borough  \
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   

                                         Neighborhood   Latitude  Longitude  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                      L'Amoreaux West, Steeles West  43.799525 -79.318389

In [76]:
print(Scarborough_venues.shape)
Scarborough_venues.head()

(84, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

## Next, we group the venues by neighborhood.

In [77]:
Scarborough_venues.groupby('Neighborhood').head()

Neighborhood  Neighborhood Latitude  \
0                                      Rouge, Malvern              43.806686   
1              Highland Creek, Rouge Hill, Port Union              43.784535   
2                   Guildwood, Morningside, West Hill              43.763573   
3                   Guildwood, Morningside, West Hill              43.763573   
4                   Guildwood, Morningside, West Hill              43.763573   
5                   Guildwood, Morningside, West Hill              43.763573   
6                   Guildwood, Morningside, West Hill              43.763573   
9                                              Woburn              43.770992   
10                                             Woburn              43.770992   
11                                             Woburn              43.770992   
12                                             Woburn              43.770992   
13                                          Cedarbrae              43.773136   
14                                          Cedarbrae              43.773136   
15                                          Cedarbrae              43.773136   
16                                          Cedarbrae              43.773136   
17                                          Cedarbrae              43.773136   
20                                Scarborough Village              43.744734   
21                                Scarborough Village              43.744734   
22        East Birchmount Park, Ionview, Kennedy Park              43.727929   
23        East Birchmount Park, Ionview, Kennedy Park              43.727929   
24        East Birchmount Park, Ionview, Kennedy Park              43.727929   
25        East Birchmount Park, Ionview, Kennedy Park              43.727929   
26        East Birchmount Park, Ionview, Kennedy Park              43.727929   
29                    Clairlea, Golden Mile, Oakridge              43.711112   
30                    Clairlea, Golden Mile, Oakridge              43.711112   
31                    Clairlea, Golden Mile, Oakridge              43.711112   
32                    Clairlea, Golden Mile, Oakridge              43.711112   
33                    Clairlea, Golden Mile, Oakridge              43.711112   
39    Cliffcrest, Cliffside, Scarborough Village West              43.716316   
40    Cliffcrest, Cliffside, Scarborough Village West              43.716316   
41    Cliffcrest, Cliffside, Scarborough Village West              43.716316   
42                        Birch Cliff, Cliffside West              43.692657   
43                        Birch Cliff, Cliffside West              43.692657   
44                        Birch Cliff, Cliffside West              43.692657   
45                        Birch Cliff, Cliffside West              43.692657   
46  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
47  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
48  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
49  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
50  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
53                                  Maryvale, Wexford              43.750072   
54                                  Maryvale, Wexford              43.750072   
55                                  Maryvale, Wexford              43.750072   
56                                  Maryvale, Wexford              43.750072   
57                                  Maryvale, Wexford              43.750072   
59                                          Agincourt              43.794200   
60                                          Agincourt              43.794200   
61                                          Agincourt              43.794200   
62                                          Agincourt              43.794200   
63            Clarks Corners, Sullivan, T

In [78]:
format(len(Scarborough_venues['Venue Category'].unique()))

'51'

## We use one hot encoding to convert the variables to be more useful in the machine learning process.

In [79]:
# one hot encoding
Scarborough_onehot = pd.get_dummies(Scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Scarborough_onehot['Neighborhood'] = Scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Scarborough_onehot.columns[-1]] + list(Scarborough_onehot.columns[:-1])
Scarborough_onehot = Scarborough_onehot[fixed_columns]

Scarborough_onehot.head()

Neighborhood  American Restaurant  \
0                          Rouge, Malvern                    0   
1  Highland Creek, Rouge Hill, Port Union                    0   
2       Guildwood, Morningside, West Hill                    0   
3       Guildwood, Morningside, West Hill                    0   
4       Guildwood, Morningside, West Hill                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    1               0   
2                   0            0       0     0    0               0   
3                   0            0       0     0    0               0   
4                   0            0       0     0    0               0   

   Bus Line  Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0         0            0     0                     0                   0   
1         0            0     0                     0                   0   
2         0            0     0                     0                   0   
3         0            0     0                     0                   0   
4         0            0     0                     0                   0   

   Clothing Store  Coffee Shop  College Stadium  Convenience Store  \
0               0            0                0                  0   
1               0            0                0                  0   
2               0            0                0                  0   
3               0            0                0                  0   
4               0            0                0                  0   

   Department Store  Discount Store  Electronics Store  Fast Food Restaurant  \
0                 0               0                  0                     1   
1                 0               0                  0                     0   
2                 0               0                  0                     0   
3                 0               0                  1                     0   
4                 0               0                  0                     0   

   Fried Chicken Joint  General Entertainment  Grocery Store  \
0                    0                      0              0   
1                    0                      0              0   
2                    0                      0              0   
3                    0                      0              0   
4                    0                      0              0   

   Hakka Restaurant  Indian Restaurant  Insurance Office  Intersection  \
0                 0                  0                 0             0   
1                 0                  0                 0             0   
2                 0                  0                 0             0   
3                 0                  0                 0             0   
4                 0                  0                 0             0   

   Italian Restaurant  Korean Restaurant  Latin American Restaurant  \
0                   0                  0                          0   
1                   0                  0                          0   
2                   0                  0                          0   
3                   0                  0                          0   
4                   0                  0                          0   

   Light Rail Station  Lounge  Medical Center  Metro Station  \
0                   0       0               0              0   
1                   0       0               0              0   
2                   0       0               0              0   
3                   0       0               0              0   
4                   0       0               0              0   

   Mexican Restaurant  Middle Eastern Restaurant  Motel  Noodle House  Park  \
0                   0                          0      0             0     0   
1                   0             

In [80]:
Scarborough_onehot.shape

(84, 52)

In [81]:
Scarborough_grouped = Scarborough_onehot.groupby('Neighborhood').mean().reset_index()
Scarborough_grouped

Neighborhood  American Restaurant  \
0                                           Agincourt             0.000000   
1   Agincourt North, L'Amoreaux East, Milliken, St...             0.000000   
2                         Birch Cliff, Cliffside West             0.000000   
3                                           Cedarbrae             0.000000   
4                     Clairlea, Golden Mile, Oakridge             0.000000   
5             Clarks Corners, Sullivan, Tam O'Shanter             0.000000   
6     Cliffcrest, Cliffside, Scarborough Village West             0.333333   
7   Dorset Park, Scarborough Town Centre, Wexford ...             0.000000   
8         East Birchmount Park, Ionview, Kennedy Park             0.000000   
9                   Guildwood, Morningside, West Hill             0.000000   
10             Highland Creek, Rouge Hill, Port Union             0.000000   
11                      L'Amoreaux West, Steeles West             0.000000   
12                                  Maryvale, Wexford             0.000000   
13                                     Rouge, Malvern             0.000000   
14                                Scarborough Village             0.000000   
15                                             Woburn             0.000000   

    Athletics & Sports  Auto Garage    Bakery      Bank  Bar  Breakfast Spot  \
0             0.000000     0.000000  0.000000  0.000000  0.0        0.250000   
1             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
2             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
3             0.142857     0.000000  0.142857  0.142857  0.0        0.000000   
4             0.000000     0.000000  0.200000  0.000000  0.0        0.000000   
5             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
6             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
7             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
8             0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
9             0.000000     0.000000  0.000000  0.000000  0.0        0.142857   
10            0.000000     0.000000  0.000000  0.000000  1.0        0.000000   
11            0.000000     0.000000  0.000000  0.000000  0.0        0.100000   
12            0.000000     0.166667  0.166667  0.000000  0.0        0.166667   
13            0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
14            0.000000     0.000000  0.000000  0.000000  0.0        0.000000   
15            0.000000     0.000000  0.000000  0.000000  0.0        0.000000   

    Bus Line  Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0        0.0     0.000000  0.00              0.000000            0.000000   
1        0.0     0.000000  0.00              0.000000            0.000000   
2        0.0     0.000000  0.25              0.000000            0.000000   
3        0.0     0.000000  0.00              0.142857            0.000000   
4        0.2     0.100000  0.00              0.000000            0.000000   
5        0.0     0.000000  0.00              0.000000            0.111111   
6        0.0     0.000000  0.00              0.000000            0.000000   
7        0.0     0.000000  0.00              0.000000            0.142857   
8        0.0     0.142857  0.00              0.000000            0.142857   
9        0.0     0.000000  0.00              0.000000            0.000000   
10       0.0     0.000000  0.00              0.000000            0.000000   
11       0.0     0.000000  0.00              0.000000            0.200000   
12       0.0     0.000000  0.00              0.000000            0.000000   
13       0.0     0.000000  0.00              0.000000            0.000000   
14       0.0     0.000000  0.00              0.000000            0.000000   
15       0.0     0.000000  0.00              0.000000            0.000000   

    Clothing Store  Coffee Shop  College St

In [82]:
Scarborough_grouped.shape

(16, 52)

## Next, we review which are the top five venues.

In [83]:
num_top_venues = 5

for hood in Scarborough_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Scarborough_grouped[Scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                 venue  freq
0               Lounge  0.25
1         Skating Rink  0.25
2       Breakfast Spot  0.25
3       Clothing Store  0.25
4  American Restaurant  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                       venue  freq
0                       Park   0.5
1                 Playground   0.5
2        American Restaurant   0.0
3          Korean Restaurant   0.0
4  Latin American Restaurant   0.0


----Birch Cliff, Cliffside West----
                   venue  freq
0           Skating Rink  0.25
1  General Entertainment  0.25
2                   Café  0.25
3        College Stadium  0.25
4    American Restaurant  0.00


----Cedarbrae----
                venue  freq
0     Thai Restaurant  0.14
1              Bakery  0.14
2                Bank  0.14
3  Athletics & Sports  0.14
4    Hakka Restaurant  0.14


----Clairlea, Golden Mile, Oakridge----
          venue  freq
0        Bakery   0.2
1      Bus Line   0.2
2          

## We then use a function to sort the most common venues.

In [84]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)    
    return row_categories_sorted.index.values[0:num_top_venues]

In [85]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Scarborough_grouped['Neighborhood']

for ind in np.arange(Scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  1st Most Common Venue  \
0                                           Agincourt         Clothing Store   
1   Agincourt North, L'Amoreaux East, Milliken, St...             Playground   
2                         Birch Cliff, Cliffside West  General Entertainment   
3                                           Cedarbrae        Thai Restaurant   
4                     Clairlea, Golden Mile, Oakridge                 Bakery   
5             Clarks Corners, Sullivan, Tam O'Shanter            Pizza Place   
6     Cliffcrest, Cliffside, Scarborough Village West                  Motel   
7   Dorset Park, Scarborough Town Centre, Wexford ...      Indian Restaurant   
8         East Birchmount Park, Ionview, Kennedy Park         Discount Store   
9                   Guildwood, Morningside, West Hill     Mexican Restaurant   
10             Highland Creek, Rouge Hill, Port Union                    Bar   
11                      L'Amoreaux West, Steeles West   Fast Food Restaurant   
12                                  Maryvale, Wexford            Auto Garage   
13                                     Rouge, Malvern   Fast Food Restaurant   
14                                Scarborough Village             Playground   
15                                             Woburn            Coffee Shop   

    2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0            Skating Rink             Breakfast Spot                Lounge   
1                    Park      Vietnamese Restaurant    Chinese Restaurant   
2            Skating Rink                       Café       College Stadium   
3      Athletics & Sports           Hakka Restaurant                Bakery   
4                Bus Line               Soccer Field          Intersection   
5     Fried Chicken Joint            Thai Restaurant  Fast Food Restaurant   
6     American Restaurant          Indian Restaurant         Grocery Store   
7               Pet Store  Latin American Restaurant    Light Rail Station   
8      Chinese Restaurant           Department Store           Bus Station   
9                     Spa          Electronics Store   Rental Car Location   
10  Vietnamese Restaurant             Clothing Store         Grocery Store   
11     Chinese Restaurant              Grocery Store        Sandwich Place   
12                 Bakery              Shopping Mall        Sandwich Place   
13  Vietnamese Restaurant             Clothing Store         Grocery Store   
14      Convenience Store      Vietnamese Restaurant        Clothing Store   
15       Insurance Office          Korean Restaurant     Indian Restaurant   

    5th Most Common Venue      6th Most Common Venue  7th Most Common Venue  \
0   Vietnamese Restaurant      General Entertainment    Fried Chicken Joint   
1   General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
2   Vietnamese Restaurant             Clothing Store          Grocery Store   
3                    Bank        Fried Chicken Joint   Caribbean Restaurant   
4    Fast Food Restaurant                Bus Station                   Park   
5      Italian Restaurant                   Pharmacy           Noodle House   
6   General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
7   Vietnamese Restaurant         Chinese Restaurant                    Bar   
8       Convenience Store                Coffee Shop         Clothing Store   
9             Pizza Place             Breakfast Spot         Medical Center   
10  General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
11         Breakfast Spot                Pizza Place               Pharmacy   
12         Breakfast Spot  Middle Eastern Restaurant  Vietnamese Restaurant   
13  General Entertainment        Fried Chicken Joint      Electronics Store   
14          Grocery Store      General Entertainment    Fried Chicken Joint   
15          Grocery Store      General Entertainment    Fried Chicken Joint   

    8th M

## K-means is used for machine learning.

In [86]:
# set number of clusters
kclusters = 3

Scarborough_grouped_clustering = Scarborough_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Scarborough_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [87]:
Scarborough_merged = Scarborough_data

# add clustering labels
Scarborough_merged['Cluster Labels'] = kmeans.labels_
# make the column name the same
Scarborough_merged.rename(columns={'Neighbourhood':'Neighborhood'}, inplace = True)

## We then merge the neighborhoods with the corresponding most common venues for each.

In [88]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Scarborough_merged = Scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Scarborough_merged.head() # check the last columns!

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0 -79.194353               0  Fast Food Restaurant  Vietnamese Restaurant   
1 -79.160497               0                   Bar  Vietnamese Restaurant   
2 -79.188711               0    Mexican Restaurant                    Spa   
3 -79.216917               0           Coffee Shop       Insurance Office   
4 -79.239476               0       Thai Restaurant     Athletics & Sports   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0        Clothing Store         Grocery Store  General Entertainment   
1        Clothing Store         Grocery Store  General Entertainment   
2     Electronics Store   Rental Car Location            Pizza Place   
3     Korean Restaurant     Indian Restaurant          Grocery Store   
4      Hakka Restaurant                Bakery                   Bank   

   6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0    Fried Chicken Joint     Electronics Store         Discount Store   
1    Fried Chicken Joint  Fast Food Restaurant      Electronics Store   
2         Breakfast Spot        Medical Center  Vietnamese Restaurant   
3  General Entertainment   Fried Chicken Joint   Fast Food Restaurant   
4    Fried Chicken Joint  Caribbean Restaurant        College Stadium   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store      Convenience Store  
1        Discount Store       Department Store  
2   Fried Chicken Joint   Fast Food Restaurant  
3     Electronics Store         Discount Store  
4         Grocery Store  General Entertainment

## We create a map to visualize the clusters.

In [89]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Scarborough_merged['Latitude'], Scarborough_merged['Longitude'], Scarborough_merged['Neighborhood'], Scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Ultimately, we verify each of the three clusters.

In [90]:
Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 0, Scarborough_merged.columns[[1] + list(range(3, Scarborough_merged.shape[1]))]]

Borough   Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0   Scarborough  43.806686 -79.194353               0   Fast Food Restaurant   
1   Scarborough  43.784535 -79.160497               0                    Bar   
2   Scarborough  43.763573 -79.188711               0     Mexican Restaurant   
3   Scarborough  43.770992 -79.216917               0            Coffee Shop   
4   Scarborough  43.773136 -79.239476               0        Thai Restaurant   
5   Scarborough  43.744734 -79.239476               0             Playground   
6   Scarborough  43.727929 -79.262029               0         Discount Store   
7   Scarborough  43.711112 -79.284577               0                 Bakery   
8   Scarborough  43.716316 -79.239476               0                  Motel   
9   Scarborough  43.692657 -79.264848               0  General Entertainment   
11  Scarborough  43.750072 -79.295849               0            Auto Garage   
12  Scarborough  43.794200 -79.262029               0         Clothing Store   
14  Scarborough  43.815252 -79.284577               0             Playground   
15  Scarborough  43.799525 -79.318389               0   Fast Food Restaurant   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0   Vietnamese Restaurant         Clothing Store         Grocery Store   
1   Vietnamese Restaurant         Clothing Store         Grocery Store   
2                     Spa      Electronics Store   Rental Car Location   
3        Insurance Office      Korean Restaurant     Indian Restaurant   
4      Athletics & Sports       Hakka Restaurant                Bakery   
5       Convenience Store  Vietnamese Restaurant        Clothing Store   
6      Chinese Restaurant       Department Store           Bus Station   
7                Bus Line           Soccer Field          Intersection   
8     American Restaurant      Indian Restaurant         Grocery Store   
9            Skating Rink                   Café       College Stadium   
11                 Bakery          Shopping Mall        Sandwich Place   
12           Skating Rink         Breakfast Spot                Lounge   
14                   Park  Vietnamese Restaurant    Chinese Restaurant   
15     Chinese Restaurant          Grocery Store        Sandwich Place   

    5th Most Common Venue      6th Most Common Venue  7th Most Common Venue  \
0   General Entertainment        Fried Chicken Joint      Electronics Store   
1   General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
2             Pizza Place             Breakfast Spot         Medical Center   
3           Grocery Store      General Entertainment    Fried Chicken Joint   
4                    Bank        Fried Chicken Joint   Caribbean Restaurant   
5           Grocery Store      General Entertainment    Fried Chicken Joint   
6       Convenience Store                Coffee Shop         Clothing Store   
7    Fast Food Restaurant                Bus Station                   Park   
8   General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
9   Vietnamese Restaurant             Clothing Store          Grocery Store   
11         Breakfast Spot  Middle Eastern Restaurant  Vietnamese Restaurant   
12  Vietnamese Restaurant      General Entertainment    Fried Chicken Joint   
14  General Entertainment        Fried Chicken Joint   Fast Food Restaurant   
15         Breakfast Spot                Pizza Place               Pharmacy   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0          Discount Store       Department Store      Convenience Store  
1       Electronics Store         Discount Store       Department Store  
2   Vietnamese Restaurant    Fried Chicken Joint   Fast Food Restaurant  
3    Fast Food Restaurant      Electronics Store         Discount Store  
4         College Stadium          Grocery Store  General Entertainment  
5    Fast Food Restaurant      Electronics Store         Discount 

In [91]:
Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 1, Scarborough_merged.columns[[1] + list(range(3, Scarborough_merged.shape[1]))]]

Borough  Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
10  Scarborough  43.75741 -79.273304               1     Indian Restaurant   

   2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
10             Pet Store  Latin American Restaurant    Light Rail Station   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10  Vietnamese Restaurant    Chinese Restaurant                   Bar   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
10  General Entertainment   Fried Chicken Joint   Fast Food Restaurant

In [92]:

Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 2, Scarborough_merged.columns[[1] + list(range(3, Scarborough_merged.shape[1]))]]

Borough   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
13  Scarborough  43.781638 -79.304302               2           Pizza Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13   Fried Chicken Joint       Thai Restaurant  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13    Italian Restaurant              Pharmacy          Noodle House   

   8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
13    Chinese Restaurant  Vietnamese Restaurant         Clothing Store

## THE END. THANK YOU!